src/golioth_block_processor.c

```diff
@@ -0,0 +1,202 @@
+    stats->block_max_ms = max(stats->block_max_ms, block_latency_ms);
+}
+
+static void download_init(
+        download_ctx_t* ctx,
+        golioth_client_t client,
+        const golioth_ota_component_t* ota_component,
+        uint8_t* download_buf) {
+    memset(ctx, 0, sizeof(*ctx));
+    block_stats_init(&ctx->block_stats);
+    ctx->ota_component = ota_component;
+    ctx->client = client;
+
+    assert(download_buf);
+    ctx->download_buf = download_buf;
+
+    // Note: total_num_blocks is an estimate of the number of blocks required to download,
+    // based on the size of the component reported in the manifest.
+    //
+    // Due to the way this size is populated in the manifest on the server, the actual
+    // number of blocks may be different.
+    ctx->total_num_blocks = golioth_ota_size_to_nblocks(ota_component->size);
+}
+
+static golioth_status_t download_block(download_ctx_t* ctx) {
+    if (ctx->is_last_block) {
+        // We've already downloaded the last block
+        return GOLIOTH_ERR_NO_MORE_DATA;
+    }
+
+    const uint64_t download_start_ms = golioth_sys_now_ms();
+
+    GLTH_LOGI(
+            TAG,
+            "Downloading block index %" PRIu32 " (%" PRIu32 "/%" PRIu32 ")",
+            (uint32_t)ctx->block_index,
+            (uint32_t)ctx->block_index + 1,
+            (uint32_t)ctx->total_num_blocks);
+
+    GOLIOTH_STATUS_RETURN_IF_ERROR(golioth_ota_get_block_sync(
+            ctx->client,
+            ctx->ota_component->package,
+            ctx->ota_component->version,
+            ctx->block_index,
+            ctx->download_buf,
+            &ctx->block_bytes_downloaded,
+            &ctx->is_last_block,
+            GOLIOTH_WAIT_FOREVER));
+
+    assert(ctx->block_bytes_downloaded <= GOLIOTH_OTA_BLOCKSIZE);
+    block_stats_update(&ctx->block_stats, golioth_sys_now_ms() - download_start_ms);
+
+    ctx->block_index++;
+
+    assert(ctx->output_fn);
+    return ctx->output_fn(ctx->download_buf, ctx->block_bytes_downloaded, ctx->output_fn_arg);
+}
+
+static void decompress_init(decompress_ctx_t* ctx, bool enable_decompression) {
+    memset(ctx, 0, sizeof(*ctx));
+    if (enable_decompression) {
+        GLTH_LOGI(TAG, "Compressed image detected");
+        ctx->hsd = heatshrink_decoder_alloc(
+                GOLIOTH_OTA_BLOCKSIZE,  // input_buffer_size
+                HEATSHRINK_WINDOW_SZ2,
+                HEATSHRINK_LOOKAHEAD_SZ2);
+    }
+}
+
+static void handle_block_init(handle_block_ctx_t* ctx, size_t component_size) {
+    memset(ctx, 0, sizeof(*ctx));
+    ctx->component_size = component_size;
+}
+
+static golioth_status_t handle_block(const uint8_t* in_data, size_t in_data_size, void* arg) {
+    handle_block_ctx_t* ctx = (handle_block_ctx_t*)arg;
+
+    golioth_status_t status = fw_update_handle_block(
+            in_data,
+            in_data_size,
+            ctx->bytes_handled,  // offset
+            ctx->component_size);
```

Comment:
This component (`golioth_block_processor.c`) is still FW related in this part.
Because of that, I would rather rename it to `golioth_fw_block_processor.c`.